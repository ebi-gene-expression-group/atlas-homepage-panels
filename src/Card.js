import React from 'react'
import PropTypes from 'prop-types'
import logo from './assets/img/HCA-logo.jpg'

class Card extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
  	var {host, cardName} = this.props
  	return (
		<div className="small-4 columns combo text-center padding-top-medium">
	  		<span>I am a card for {cardName}</span> <br />
	        <a href={`${host}/${cardName.toLowerCase()}`}>
	            <img className="margin-bottom-large experiment-thumbnail" src={logo}  height="800" />
	            <h6>
	                <a href="/gxa/experiments/E-MTAB-4344">Human tissues</a><br />
	                <a href="/gxa/experiments/E-GEOD-26284">Human cells</a>
	            </h6>
	        </a>
		</div>
  	)

  }
}

Card.propTypes = {
  host: PropTypes.string.isRequired,
  CardType: PropTypes.func
}

export default Card
