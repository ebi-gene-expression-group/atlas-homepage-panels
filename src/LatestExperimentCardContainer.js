import React from 'react'
import PropTypes from 'prop-types'
import LatestExperimentCard from './LatestExperimentCard'
import URI from 'urijs'

const CalloutAlert = ({error}) =>
  <div className={`row column`}>
    <div className={`callout alert small`}>
      <h5>Oops!</h5>
      <p>
        {error.description}<br/>
        If the error persists, in order to help us debug the issue, please copy the URL and this message and
        send it to us via <a href={`https://www.ebi.ac.uk/support/gxasc`}>the EBI Support & Feedback system</a>:
      </p>
      <code>{`${error.name}: ${error.message}`}</code>
    </div>
  </div>

CalloutAlert.propTypes = {
  error: PropTypes.shape({
    description: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  })
}

class LatestExperimentCardContainer extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      data: null,
      isLoading: true,
      hasError: null
    }
  }

  async componentDidMount() {
    this.setState({ isLoading: true })

    const url = URI(this.props.resource, this.props.host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      this.setState({
        data: await response.json(),
        isLoading: false,
        hasError: null
      })
    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { data, isLoading, hasError } = this.state

    const cards = data && data.map((card, idx) => <LatestExperimentCard key={idx} host={this.props.host} {...card}/>)

    return (
      hasError ?
        <CalloutAlert error={hasError} /> :
        isLoading ?
          <p className={`row column`} id={`loading-message`}> Loading, please wait...</p> :
          data.length > 0 ?
            <div className={`row small-up-2 medium-up-3`}>
              {cards}
            </div> :
            null
    )
  }
}

LatestExperimentCardContainer.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired
}

export default LatestExperimentCardContainer