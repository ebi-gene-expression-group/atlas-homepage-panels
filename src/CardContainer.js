import React from 'react'
import PropTypes from 'prop-types'
import Card from './Card'
import LatestExperimentCardContainer from './LatestExperimentCardContainer'

const CardContainer = ({panelName, host, resource}) => {
  switch(panelName) {
    case `Featured`:
      return <Card host={host} resource={resource[`Featured`]}/>
      break
    case `Latest experiments`:
      return <LatestExperimentCardContainer host={host} resource={resource[`Latest`]}/>
      break
    default:
      return null
  } 
}

CardContainer.propTypes = {
  panelName: PropTypes.string.isRequired,
  host: PropTypes.string.isRequired,
  resource: PropTypes.object.isRequired
}

export default CardContainer
