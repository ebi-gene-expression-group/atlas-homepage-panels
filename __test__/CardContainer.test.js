import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {shallow, mount, render} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import CardContainer from '../src/CardContainer'
import LatestExperimentCardContainer from '../src/LatestExperimentCardContainer'
import Card from '../src/Card'

Enzyme.configure({ adapter: new Adapter() })

describe(`CardContainer`, () => {
  const props = {  host: `link`, panelName: `panel`, CardType: ()=>`I am a card`}
  test(`uses a LatestExperimentCard for latest experiment entries`, () => {
     expect(shallow(<CardContainer {...props} panelName={`Latest experiments`}/>).find(LatestExperimentCardContainer).length).toEqual(1)
     expect(shallow(<CardContainer {...props} panelName={`Latest experiments`}/>).find(Card).length).toEqual(0)
     expect(shallow(<CardContainer {...props} panelName={`Latest experiments`}/>).find(props.CardType).length).toEqual(0)
  })

  test(`uses a static card for featured experiment`, () => {
     expect(shallow(<CardContainer {...props} panelName={`Featured`}/>).find(LatestExperimentCardContainer).length).toEqual(0)
     expect(shallow(<CardContainer {...props} panelName={`Featured`}/>).find(Card).length).toEqual(1)
     expect(shallow(<CardContainer {...props} panelName={`Featured`}/>).find(props.CardType).length).toEqual(0)
  })

  test(`uses a default/prop card for species experiment`, () => {
     expect(shallow(<CardContainer {...props} />).find(LatestExperimentCardContainer).length).toEqual(0)
     expect(shallow(<CardContainer {...props} />).find(Card).length).toEqual(0)
     expect(shallow(<CardContainer {...props} />).find(props.CardType).length).toEqual(1)
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<CardContainer {...props}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
