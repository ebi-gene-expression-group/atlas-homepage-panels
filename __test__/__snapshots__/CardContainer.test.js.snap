// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer matches snapshot 1`] = `
<div
  className="row margin-bottom-xxlarge"
>
  I am a card
</div>
`;
